### PR TITLE
Support standard AWS config in the S3 remote backend.

### PR DESCRIPTION
**Pull request by johnrengelman**
Synchronizes the AWS Credential configuration code with the S3 Remote Backend. This allows the S3 remote backend to understand shared credential files and credential profiles.

### CODE REVIEW

**Comment by johnrengelman** [comment]
Would be nice for the `getCreds` function to be shared between the S3 backend and the AWS provider, but I couldn't think of a good place for it to go so that both packages could use it.

**Comment by kendawg2** [comment]
+1 for this.   Selecting a profile from a shared credentials file would be nice.   The code seems to have the hook for it, but not obvious how to pass it.    Also, not sure I agree with using the AWS provider config since some people use a common S3 bucket for all states even though they deploy using other accounts.   This is exactly the problem I am encountering.   I have to set the S3 crews in environment variable and then I can't use a profile in the AWS provider because it seems to look at the reds provided there first, then environment variables, then lastly to the shared credential file.   If you pass the credentials on the command line in the remote config it stores them in the state file, which is not good.

**Comment by johnrengelman** [comment]
@kendawg2 it's just the same code...it doesn't share the actual configuration.

**Comment by johnrengelman** [comment]
@kendawg2 the docs were update to reflect how to pass the profile: https://github.com/hashicorp/terraform/pull/5270/files#diff-51fe5c0fa7e6751718babdfccde26120R55

**Comment by kendawg2** [comment]
@johnrengelman Got it.   Saw the docs.   Thanks.

**Comment by radeksimko** [comment]
This is looking pretty good, I won't be against if someone reviews & merges this, but I personally want to focus on finishing https://github.com/hashicorp/terraform/pull/5030 or more specifically on https://github.com/aws/aws-sdk-go/issues/543 which is very likely going to have impact on this too.

Also I agree with you that the S3 remote backend logic for getting credentials should be leveraging the existing logic we have in AWS provider. I think the easiest way to achieve it would be to expose `aws.getCreds` as `aws.GetCreds` or even better `aws.GetCredentials`. Then you could import the provider here like:

```go
import "github.com/hashicorp/terraform/builtin/providers/aws"
```

**Comment by kendawg2** [comment]
Any update on this?  This actually fixes a problem I am having and it sure would be nice to have merged.

**Comment by radeksimko** [comment]
@kendawg2 I just finished refactoring https://github.com/hashicorp/terraform/pull/5030 which also now exposes `aws.GetCredentials` and `aws.GetAccountId` as discussed above.

It will need to be peer reviewed by another maintainer & merged, hopefully that will happen soon and then it's mostly up to @johnrengelman to leverage the refactored code here.

@johnrengelman you are also welcomed to have a look :eyes: at the mentioned PR to confirm it's solving the problems with duplicity here in this PR.

Finally someone (potentially myself) can review and merge this PR.

**Comment by toddrosner** [comment]
It'd be awesome to have this one ship with 0.6.15

**Comment by ashmeer7** [comment]
+1 for getting this in 0.6.15!

**Comment by johnrengelman** [comment]
I rebased this on latest master and moved to just sharing the `GetCredentials` function from the AWS provider since #5030 hasn't been merged yet. 
I'd be happy to look at something once that is in, but if it's going to be delayed, I think this should go in regardless.

**Comment by radeksimko** [comment]
Hi @johnrengelman 
https://github.com/hashicorp/terraform/pull/5030 has been merged and exposes `aws.GetCredentials` which you can now leverage here in this PR.

Would you mind refactoring this PR?

**Comment by johnrengelman** [comment]
@radeksimko done.

File: website/source/docs/state/remote/s3.html.md [hunk 1]
```diff
@@ -56,3 +56,6 @@ The following configuration options / environment variables are supported:
  * `access_key` / `AWS_ACCESS_KEY_ID` - (Optional) AWS access key
  * `secret_key` / `AWS_SECRET_ACCESS_KEY` - (Optional) AWS secret key
  * `kms_key_id` - (Optional) The ARN of a KMS Key to use for encrypting the state.
+ * `profile` - (Optional) This is the AWS profile name as set in the shared credentials file.
+ * `shared_credentials_file`  - (Optional) This is the path to the shared credentials file. If this is not set and a profile is specified, ~/.aws/credentials will be used.
+ * `token` - (Optional) Use this to set an MFA token. It can also be sourced from the `AWS_SECURITY_TOKEN` environment variable.
```
Comment (radeksimko) [inline website/source/docs/state/remote/s3.html.md]: This is not really an MFA token, it's session token and can be used any time user uses `AssumeRole*` API - no matter if it's MFA-protected or isn't, but that's really a nitpick, I can fix that in a separate commit.

**Comment by sammysammy** [comment]
Not sure if this helps you but just discovered putting the profile in the backend config works - like so (in 0.7rc2 - haven't tried with 16 yet):

terraform remote config -backend=s3 -backend-config="bucket=<yourbucketname>" -backend-config="key=testing" -backend-config="region=us-west-2" -backend-config="profile=<aws-profile-name>"

**Pull request merged**